src/cas_cli/client.py

```diff
@@ -0,0 +1,238 @@
+        feature_ids_column_name: str,
+    ) -> anndata.AnnData:
+        try:
+            cas_feature_schema_list = self._feature_schemas_cache[feature_schema_name]
+        except KeyError:
+            cas_feature_schema_list = self.cas_api_service.get_cas_pca_002_schema_from_dump()
+            self._feature_schemas_cache[feature_schema_name] = cas_feature_schema_list
+
+        try:
+            data_preparation.validate(
+                adata=adata,
+                cas_feature_schema_list=cas_feature_schema_list,
+                feature_ids_column_name=feature_ids_column_name,
+            )
+        except exceptions.DataValidationError as e:
+            if e.extra_features > 0:
+                self._print(
+                    f"Input datafile has {e.extra_features} more features than {feature_schema_name} schema in CAS. "
+                    f"They will be omitted..."
+                )
+            if e.missing_features > 0:
+                self._print(
+                    f"Input datafile has {e.missing_features} missing features compared with {feature_schema_name} in CAS. "
+                    f"They will be filled with zeros..."
+                )
+            if e.extra_features == 0 and e.missing_features == 0:
+                self._print(
+                    f"Input datafile has all the necessary features as {feature_schema_name}, but it's still "
+                    f"incompatible because of the different order. The features will be reordered "
+                    f"according to {feature_schema_name}..."
+                )
+            return data_preparation.sanitize(
+                adata=adata,
+                cas_feature_schema_list=cas_feature_schema_list,
+                count_matrix_name=count_matrix_name,
+                feature_ids_column_name=feature_ids_column_name,
+            )
+        else:
+            self._print(f"Input dataset is validated and fully compatible with {feature_schema_name} schema...")
+            return adata
+
+    async def _annotate_anndata_chunk(
+        self,
+        adata_bytes: bytes,
+        results: t.List,
+        chunk_index: int,
+        chunk_start_i: int,
+        chunk_end_i: int,
+    ) -> None:
+        """
+        A wrapper around POST request that handles HTTP 500 and HTTP 401 status responses
+        In case of HTTP 500 response resubmit task 2 times.
+        In case of HTTP 501 print a message
+
+        :param adata_bytes: Dumped bytes of anndata chunk
+        :param results: Results list that needs to be used to inplace the response from the server
+        :param chunk_index: Consequent number of the chunk (e.g Chunk 1, Chunk 2)
+        :param chunk_start_i: Index pointing to the main adata file start position of the current chunk
+        :param chunk_end_i: Index pointing to the main adata file end position of the current chunk
+        """
+        number_of_cells = chunk_end_i - chunk_start_i
+        for _ in range(3):
```

Comment:
3 is a magic number. Replace with `num_attempts_per_chunk` and expose to the user.

Reply:
make it a kwarg of `CASClient.__init__` with a default value of 3.